### PR TITLE
Sorting experiences

### PR DESCRIPTION
**Pull request by BayLadyCoder**
It works in the web app (frontend), but the test fails. 

The getMonthScore function test passes, but sortExperiences function test fails.

I couldn't find any problems in the code. Hope you can find something to fix the failed test.

Thank you.

### CODE REVIEW

**Comment by dfkoh** [comment]
So...the short answer to the "why is the test failing" is because your sorting code is doing exactly what it says it should do. In other words, you have a bug (or maybe more than one, depending on how you count)

Coincidentally you've given it an array that is exactly in the order it wants to sort in, but if you, for instance, reorder the items in the given array, you'll find that it re-sorts them back to the order you originally specified them (this is how I figured this out).

It's also not super clear from the test what you actually want the sorting criteria to be? This is the hazard of trying to write one big test rather than a bunch of little ones which each just test one thing (I'm guilty of this at times as well). However, given you're having trouble getting this to work, I'd adding smaller tests that you only test one thing in a given test (e.g. there's only one assertion).

For instance, you could have a test that just tests that `is_current` things are sorted before non-`is_current` things (one issue with the code as it is right now, for instance).

On top of this, the actual code you've written is very hard to read or reason about. I'll drop a more specific comment about that inline

File: src/modules/Experiences/ExperiencesList/helpers.js [hunk 1]
```diff
@@ -84,4 +84,61 @@ const configureForm = (expType) => {
...
+      if (exp1.start_year < exp2.start_year) return 1;
+      if (exp1.start_month_score > exp2.start_month_score) return -1;
+      if (exp1.start_month_score < exp2.start_month_score) return 1;
+    } else if (exp1.is_current === false && exp2.is_current === false) {
+      if (exp1.end_year > exp2.end_year) return -1;
+      if (exp1.end_year < exp2.end_year) return 1;
+      if (exp1.end_month_score > exp2.end_month_score) return -1;
+      if (exp1.end_month_score < exp2.end_month_score) return 1;
+    }
+    if (exp1.is_current === true) return -1;
```
Comment (dfkoh) [inline src/modules/Experiences/ExperiencesList/helpers.js]: This code is really hard to follow. Some tactical suggestions:

Specifically when writing a sort key function, it's easiest if the logic follows a consistent pattern. I think here you have a pretty simple pattern, called a `lexicographic ordering` (basically the way you alphabetize words). It means you first compare the first thing, and sort by that, then if the first thing is equal you move onto the second thing.

Based on your test, it looks like the ordering you want is this:

1. is_current (true goes above false)
2. end_year
3. end_month
4. start_year
5. start_month

You can then write some predictable sort code by using the pattern:
```javascript
if (exp1.is_current !== exp2.is_current) {
 // return ...
} else if (exp1.end_year !== exp2.start_year) {
 // etc...
```

File: src/modules/Experiences/ExperiencesList/helpers.js [hunk 2]
```diff
@@ -84,4 +84,61 @@ const configureForm = (expType) => {
...
+};
+
+const sortExperiences = (experiences) => {
+  const experiencesWithScores = getMonthScore(experiences);
+
+  const sortedExperiences = experiencesWithScores.sort(function(exp1, exp2) {
+    // if is_current is True or have the same end month/year
+    // sort from start month/year (newest experience)
+    if (exp1.end_month === exp2.end_month && exp1.end_year === exp2.end_year) {
+      if (exp1.start_year > exp2.start_year) return -1;
```
Comment (dfkoh) [inline src/modules/Experiences/ExperiencesList/helpers.js]: When you're comparing numeric values, you can just do `return exp2.start_year - exp1.start_year`. Of course, that assumes you don't want to fall back to another comparison if they are equal

File: src/modules/Experiences/ExperiencesList/helpers.js [hunk 3]
```diff
@@ -84,4 +84,61 @@ const configureForm = (expType) => {
...
+      if (exp1.start_month_score > exp2.start_month_score) return -1;
+      if (exp1.start_month_score < exp2.start_month_score) return 1;
+    } else if (exp1.is_current === false && exp2.is_current === false) {
+      if (exp1.end_year > exp2.end_year) return -1;
+      if (exp1.end_year < exp2.end_year) return 1;
+      if (exp1.end_month_score > exp2.end_month_score) return -1;
+      if (exp1.end_month_score < exp2.end_month_score) return 1;
+    }
+    if (exp1.is_current === true) return -1;
+  });
```
Comment (dfkoh) [inline src/modules/Experiences/ExperiencesList/helpers.js]: Also, one obvious bug here is that you can hit the end of the function without returning a number. You should always ensure your functions end with an unconditional `return` (or an `if/else` with a return in both branches)

File: src/modules/Experiences/ExperiencesList/helpers.js (hunk 2 above)
Comment (BayLadyCoder) [reply src/modules/Experiences/ExperiencesList/helpers.js]: So much better this way. Thanks.

File: src/modules/Experiences/ExperiencesList/helpers.js (hunk 1 above)
Comment (BayLadyCoder) [reply src/modules/Experiences/ExperiencesList/helpers.js]: Done.

File: src/modules/Experiences/ExperiencesList/helpers.js [hunk 4]
```diff
@@ -84,4 +84,57 @@ const configureForm = (expType) => {
...
+  const sortedExperiences = experiencesWithScores.sort(function(exp1, exp2) {
+    if (exp1.is_current !== exp2.is_current) {
+      return exp1.is_current === true ? -1 : 1;
+    } else {
+      return (
+        exp2.end_year - exp1.end_year ||
+        exp2.end_month - exp1.end_month ||
+        exp2.start_year - exp1.start_year ||
+        exp2.start_month - exp1.start_month
+      );
```
Comment (dfkoh) [inline src/modules/Experiences/ExperiencesList/helpers.js]: Code is much more readable now. Also I'd never considered using the `||` in a sorting function like this, that's very clever (and clean). Good job!

One last thing is even though the code is fairly readable now, it's probably still complicated enough that its worth adding a comment to explain what the sort order is supposed to be (e.g. sort by is_current, then end_date, then start_date) This will save other people (or your future self) time in having to parse through what the function is doing in the future.

I generally recommend doing this with any piece of code that was tricky to write, assuming that if it was tricky to write, it will also probably be tricky to read.
